linear_model.py

```diff
@@ -0,0 +1,118 @@
+
+import os
+import sys
+
+import tensorflow as tf
+
+class LinearModel:
+
+    def __init__(self, input_size, output_size, checkpoint=None):
+        self.input_size = input_size
+        self.output_size = output_size
+
+        self.sess = tf.Session()
+        self._build()
+        self.saver = tf.train.Saver()
+
+        if checkpoint is not None:
+            self.restore_model(checkpoint)
+        else:
+            self.sess.run(tf.global_variables_initializer())
+
+        self.iter_number = 0
+
+    ## Public methods
+
+    def predict(self, state):
+        """
+        state: B x I
+        """
+        output = self.sess.run(self.out, feed_dict={self.input_state: state})
+        return output
+
+    def train_reinforce(self, state, pred, action, var, reward):
+        """
+        B = batch size
+        I = input size
+        A = action dimension
+
+        state: B x I    input state being trained
+        pred: B x A     output from model.predict(state)
+        action: B x A   sampled action using mean from pred
+        var: B x A      variance used for sampling action (MUST BE POSTIIVE)
+        reward: B x 1   reward from environment for the performed action
+
+        inputs expected to be numpy arrays
+        """
+
+        # TODO: for actor-critic, pred should be predicted from a separate critic network instead
+
+        target = (action - pred) * reward / (2.0 * var)
```

Comment:
I think this should be r (action - pred)/(2.0 * var) - pred? I might have made a mistake. 
The argument is that the derivative (with respect to theta) of (pred(state) - target)^2 is 2 * (pred(state) - target) * grad(pred(state)). Train will take a step in the direction of -grad = (target - pred(state))^2 is 2 * (pred(state) - target) * grad(pred(state)). So we want 2 * (target - pred(state)) = (action - pred) * reward / var. Solving this, we get what I put above? I could be making a mistake.